bittensor/utils/registration/async_registration.py

```diff
@@ -0,0 +1,536 @@
+    subtensor: "AsyncSubtensor",
+    netuid: int,
+    old_block_number: int,
+    hotkey_bytes: bytes,
+    curr_diff: Array,
+    curr_block: Array,
+    curr_block_num: Value,
+    update_curr_block_: "Callable",
+    check_block: Lock,
+    solvers: list[Solver],
+    curr_stats: "RegistrationStatistics",
+) -> int:
+    """
+    Check for the newest block and update block-related information and states across solvers if a new block is detected.
+
+    Args:
+        subtensor (AsyncSubtensor): The subtensor instance interface.
+        netuid (int): The network UID for the blockchain.
+        old_block_number (int): The previously known block number.
+        hotkey_bytes (bytes): The bytes representation of the hotkey.
+        curr_diff (Array): The current difficulty level.
+        curr_block (Array): The current block information.
+        curr_block_num (Value): The current block number.
+        update_curr_block_ (Callable): Function to update current block information.
+        check_block (Lock): Lock object for synchronizing block checking.
+        solvers (list[Solver]): List of solvers to notify of new blocks.
+        curr_stats (RegistrationStatistics): Current registration statistics to update.
+
+    Returns:
+        int: The updated block number which is the same as the new block
+             number if it was detected, otherwise the old block number.
+    """
+    block_number = await subtensor.substrate.get_block_number(None)
+    if block_number != old_block_number:
+        old_block_number = block_number
+        # update block information
+        block_number, difficulty, block_hash = await _get_block_with_retry(
+            subtensor=subtensor, netuid=netuid
+        )
+        block_bytes = bytes.fromhex(block_hash[2:])
+
+        update_curr_block_(
+            curr_diff,
+            curr_block,
+            curr_block_num,
+            block_number,
+            block_bytes,
+            difficulty,
+            hotkey_bytes,
+            check_block,
+        )
+        # Set new block events for each solver
+
+        for worker in solvers:
+            worker.newBlockEvent.set()
+
+        # update stats
+        curr_stats.block_number = block_number
+        curr_stats.block_hash = block_hash
+        curr_stats.difficulty = difficulty
+
+    return old_block_number
+
+
+async def _block_solver(
+    subtensor: "AsyncSubtensor",
+    wallet: "Wallet",
+    num_processes: int,
+    netuid: int,
+    dev_id: list[int],
+    tpb: int,
+    update_interval: int,
+    curr_block,
+    curr_block_num,
+    curr_diff,
+    n_samples,
+    alpha_,
+    output_in_place,
+    log_verbose,
+    cuda: bool,
+):
+    """Shared code used by the Solvers to solve the POW solution."""
+    limit = int(math.pow(2, 256)) - 1
+
+    # Establish communication queues
+    # See the _Solver class for more information on the queues.
+    stop_event = Event()
+    stop_event.clear()
+
```

Comment:
```suggestion
    if cuda:
        num_processes = len(dev_id)
```

You need to set the num_processes here if cuda to ensure the `finished_queues = ` line will execute properly. Otherwise you'll be passing `None` as an arg to `range()`